File: apps/auth-server/src/main/java/com/example/chat/auth/server/infrastructure/persistence/RefreshTokenEntity.java
package com.example.chat.auth.server.infrastructure.persistence;

import com.example.chat.auth.server.core.domain.RefreshToken;
import com.example.chat.auth.server.core.domain.credential.Device;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "auth_refresh_tokens")
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class RefreshTokenEntity {
    @Id
    private UUID id;

    @Column(name = "principal_id", nullable = false)
    private UUID principalId;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "deviceId", column = @Column(name = "device_id")),
            @AttributeOverride(name = "platform", column = @Column(name = "platform")),
            @AttributeOverride(name = "browser", column = @Column(name = "browser"))
    })
    private Device device;

    @Column(name = "token_value", length = 1024, nullable = false, unique = true)
    private String tokenValue;

    @Column(name = "expiry_at", nullable = false)
    private Instant expiryAt;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "last_used_at")
    private Instant lastUsedAt;

    public RefreshToken toDomain() {
        return RefreshToken.builder()
                .id(id)
                .principalId(principalId)
                .device(device)
                .tokenValue(tokenValue)
                .expiryAt(expiryAt)
                .createdAt(createdAt)
                .lastUsedAt(lastUsedAt != null ? lastUsedAt : createdAt)
                .build();
    }

    public static RefreshTokenEntity fromDomain(RefreshToken domain) {
        return RefreshTokenEntity.builder()
                .id(domain.getId())
                .principalId(domain.getPrincipalId())
                .device(domain.getDevice())
                .tokenValue(domain.getTokenValue())
                .expiryAt(domain.getExpiryAt())
                .createdAt(domain.getCreatedAt())
                .lastUsedAt(domain.getLastUsedAt())
                .build();
    }
}

File: apps/auth-server/src/main/kotlin/com/example/chat/auth/server/infrastructure/persistence/RefreshTokenEntity.kt
package com.example.chat.auth.server.infrastructure.persistence

import com.example.chat.auth.server.core.domain.RefreshToken
import com.example.chat.auth.server.core.domain.credential.Device
import jakarta.persistence.*
import java.time.Instant
import java.util.*

@Entity
@Table(name = "auth_refresh_tokens")
class RefreshTokenEntity(
        @Id val id: UUID,
        @Column(name = "principal_id", nullable = false) val principalId: UUID,
        @Embedded
        @AttributeOverrides(
                AttributeOverride(name = "deviceId", column = Column(name = "device_id")),
                AttributeOverride(name = "platform", column = Column(name = "platform")),
                AttributeOverride(name = "browser", column = Column(name = "browser"))
        )
        val device: Device,
        @Column(name = "token_value", length = 1024, nullable = false, unique = true)
        val tokenValue: String,
        @Column(name = "expiry_at", nullable = false) val expiryAt: Instant,
        @Column(name = "created_at", nullable = false) val createdAt: Instant,
        @Column(name = "last_used_at") val lastUsedAt: Instant?
) {
    fun toDomain(): RefreshToken {
        return RefreshToken(
                id = id,
                principalId = principalId,
                device = device,
                tokenValue = tokenValue,
                expiryAt = expiryAt,
                createdAt = createdAt,
                lastUsedAt = lastUsedAt ?: createdAt
        )
    }

    companion object {
        fun fromDomain(domain: RefreshToken): RefreshTokenEntity {
            return RefreshTokenEntity(
                    id = domain.id,
                    principalId = domain.principalId,
                    device = domain.device,
                    tokenValue = domain.tokenValue,
                    expiryAt = domain.expiryAt,
                    createdAt = domain.createdAt,
                    lastUsedAt = domain.lastUsedAt
            )
        }
    }
}
